ADBDEV-4799: Fix the environment for running test suite for gpdb7

KnightMurloc opened the pull request:
Fix the environment for running test suite for gpdb7

The container base has been changed from CentOS 7 to Rocky 8.8.
Dnf is now used instead of yum as a package manager.
Devtoolset-7 was removed since the necessary tools are available in the base
repository.
RSA1 key generation was removed since it's deprecated.
In docker-compose, the coordinator's name has been changed from mdw to cdw.
Containers also now use the init process to prevent zombie processes from
appearing.
Packages for llvm and clang are taken from the vault, since only clang-16 is
available in the main repository, and the build gpdb with llvm 16+ is not
supported.
Cherry-picked commit c24d7e60d7f8569522b2716b77e0905c91ad3d4f to fix gpfdist
tests with TLS 1.0 and TLS 1.1.
The attachment of gpdb logs to the allure report will be added as a separate patch.

Commit:
Fix gpfdsits old ssl test cases for gp7 (#16370)

Gpfdsits old ssl test cases are used for testing TLS1.0 and TLS1.1.
We have extracted these test cases to a standalone file in previous PR #16277,
but missed to start gpfdist for these test cases.
So, add the steps to start/stop gpfdist

cherry-picked from c24d7e60d7f8569522b2716b77e0905c91ad3d4f

RomaZe committed:
Upgrade Java 1.8 to Java 11 in Dockerfile

We are going to move our programs to be running on Java 11. All these programs
use the docker image (arenadata/Dockerfile) to run a lot of different kinds of
tests. So, we need Java 11 to test these programs. This patch replaces Java 1.8
with Java 11.

cherry-picked from c097366900c7699f6e10fe5f29a763792d34af41

andr-sokolov commented on arenadata/Dockerfile:
```diff
@@ -1,45 +1,55 @@
+    dnf -y install libevent-devel libuuid-devel libxml2-devel libyaml-devel libzstd-devel && \
+    dnf -y install openssl-devel pam-devel readline-devel snappy-devel libuv-devel && \
+    dnf -y install apache-ivy libicu perl-ExtUtils-Embed perl-Env perl-JSON && \
+    dnf -y install perl-IPC-Run perl-Test-Base libxslt-devel openldap-devel libpq-devel && \
```
Do we really need libpq-devel?

KnightMurloc replied:
packages from the repository are installed in the upstream.
https://github.com/greenplum-db/gpdb/blob/e7c2b1f14bb42a1018ac57d14f4436880e0a0515/README.Rhel-Rocky.bash#L5-L12

andr-sokolov commented on arenadata/Dockerfile:
```diff
@@ -1,45 +1,55 @@
+RUN dnf -y install 'dnf-command(config-manager)' && \
+    dnf config-manager --set-enabled devel && dnf makecache && dnf update -y ca-certificates && \
+    rpm --import https://mirror.yandex.ru/rockylinux/RPM-GPG-KEY-Rocky-8 && \
+    dnf -y install epel-release java-1.8.0-openjdk-devel && \
```
Update java version
https://github.com/arenadata/gpdb/commit/c097366900c7699f6e10fe5f29a763792d34af41

andr-sokolov commented on arenadata/Dockerfile:
```diff
@@ -1,45 +1,55 @@
+    dnf -y install perl-IPC-Run perl-Test-Base libxslt-devel openldap-devel libpq-devel && \
+    dnf clean all
+
+RUN rpm --import https://dl.rockylinux.org/pub/rocky/RPM-GPG-KEY-Rocky-8 && \
```
Are RPM-GPG-KEY-Rocky-8 here and above the same files?

KnightMurloc replied:
they are the same, so this line can be omitted.

andr-sokolov commented on arenadata/Dockerfile:
```diff
@@ -1,45 +1,55 @@
+                llvm-static-15.0.7-1.module+el8.8.0+1144+0a4e73bd \
+                llvm-devel-15.0.7-1.module+el8.8.0+1144+0a4e73bd && \
+    dnf install -y \
+                --repofrompath=AppStream_external,https://dl.rockylinux.org/vault/rocky/8.8/AppStream/x86_64/os \
```
Why we have two repositories to install llvm and clang?

KnightMurloc replied:
yes, we can install all the necessary packages from one repository.

andr-sokolov commented on arenadata/Dockerfile:
```diff
@@ -1,45 +1,55 @@
-    ssh-keygen -f /etc/ssh/ssh_host_key -N '' -t rsa1 && \
     ssh-keygen -f /etc/ssh/ssh_host_rsa_key -N '' -t rsa && \
     ssh-keygen -f /etc/ssh/ssh_host_dsa_key -N '' -t dsa && \
+    /sbin/sshd & sleep 5 && \
```
Is it possible to decrease changes here?

andr-sokolov commented on arenadata/Dockerfile:
```diff
@@ -1,45 +1,55 @@
+                --repofrompath=AppStream_external,https://dl.rockylinux.org/vault/rocky/8.8/AppStream/x86_64/os \
+                clang-resource-filesystem-15.0.7-1.module+el8.8.0+1144+0a4e73bd \
+                clang-libs-15.0.7-1.module+el8.8.0+1144+0a4e73bd \
+                clang-15.0.7-1.module+el8.8.0+1144+0a4e73bd && \
```
How many compilers is it need to build gpdb?

KnightMurloc replied:
clang is required for an llvm-enabled build.
```
configure: error: clang not found, but required when compiling --with-llvm, specify with CLANG=
```

KnightMurloc replied:
clang pulls gcc as a dependency.

Stolb27 commented on arenadata/Dockerfile:
```diff
@@ -1,45 +1,55 @@
+    dnf -y install git iproute net-tools openssh-server rsync sudo time vim wget unzip \
+                   ant-junit autoconf bison cmake3 flex gperftools indent jq libtool gcc-c++ \
+                   krb5-server krb5-workstation xerces-c-devel file netcat passwd lsof xz && \
     yum clean all
```
It's the last and only presence of yum in this file. Is it correct?

Stolb27 reviewed (commented):
Please don't merge this PR. I think we may test CI on it.

KnightMurloc committed:
Fixed docker files for running the gpdb7 test suite

The container base has been changed from CentOS 7 to Rocky 8.8.
Dnf is now used instead of yum as a package manager.
Devtoolset-7 was removed since the necessary tools are available in the main
repository.
Added installation of llvm and clang for building with JIT support.
Packages for llvm and clang are taken from the vault, since only clang-16 is
available in the main repository, and the build gpdb with llvm 16+ is not
supported.
RSA1 key generation was removed since it's deprecated.
The mdw container was renamed to cdw.
Now, instead of a combination of bash and tail, the init process and sleep
infinity are used in containers to prevent zombie processes from appearing.
In the README updated the docker commands and removed outdated information.

RekGRpth commented:
cherry-picked commits lost authorship

RekGRpth commented on src/bin/gpfdist/regress/output/gpfdist_old_ssl.source:
```diff
@@ -1,3 +1,37 @@
+drop external table if exists gpfdist_ssl_stop;
+NOTICE:  table "gpfdist_ssl_stop" does not exist, skipping
+-- end_ignore
+-- --------------------------------------
```
ignore content in output is redundant

KnightMurloc replied:
this is cherry pick. is it worth editing it?

RekGRpth replied:
> this is cherry pick

OK

bandetto commented on arenadata/Dockerfile:
```diff
@@ -1,45 +1,53 @@
     echo -e "password\npassword" | passwd 2> /dev/null && \
-    { ssh-keyscan localhost; ssh-keyscan 0.0.0.0; } >> /root/.ssh/known_hosts && \
-    #
-    ssh-keygen -f /etc/ssh/ssh_host_key -N '' -t rsa1 && \
```
Why was key generation removed?

KnightMurloc replied:
rsa1 is deprecated and an attempt to generate it fails.

bandetto commented on arenadata/docker-compose.yaml:
```diff
@@ -4,19 +4,20 @@ version: "3"
       kernel.sem: 500 1024000 200 4096
+    init: true
     entrypoint: >
-      bash -c "set -e && tail -f /home/gpadmin/gpdb_src/README.md"
```
Is this line equivalent to `sleep infinity`? Does enabling init increase overhead?

KnightMurloc replied:
yes, it is equivalent, but in rocky 8 it does not work the way it does in centos 7. the task of the command `bash -c "set -e && tail -f /home/gpadmin/gpdb_src/README.md"` is to start an infinite wait and that the process with pid 1 is bash (for the correct completion of "orphaned" processes) but in rocky 8, pid 1 gets tail (possibly due to some new bash optimizations) and because of that, zombie processes begin to appear.
I don't think the init process will add an overhead.

bandetto commented on arenadata/Dockerfile:
```diff
@@ -1,45 +1,53 @@
-    yum -y install perl-IPC-Run perl-Test-Base libxslt-devel openldap-devel && \
-    yum clean all
+RUN dnf makecache && \
+    dnf -y install python39-devel && \
```
I tried to run tests in the container before this patch and got a error about python version. Why a different version was needed and how this issue was fixed?

KnightMurloc replied:
gpdb 7 migrated from python 2 to python 3

bandetto replied:
I meant that the package `python3-devel` contained the wrong version of Python (`python3.6`) and now the package `python39-devel` has the correct version of Python.

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1047327

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1047779

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1048016

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1048015

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1049307

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1051594

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1052385

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1052384

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1053666

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1053665

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1053663

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1053662

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1054872

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1054871

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1054868

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1054869

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1057738

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1057737

BenderArenadata commented:
Failed job Resource group isolation tests 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1062178

BenderArenadata commented:
Failed job Resource group isolation tests 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1062821

BenderArenadata commented:
Failed job Behave tests 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1062820

RekGRpth commented:
Complete the description of the PR in accordance with the requirements (title, width 80, ...)

RekGRpth commented:
Why first commit does not correspond to https://github.com/arenadata/gpdb/commit/c097366900c7699f6e10fe5f29a763792d34af41?

KnightMurloc commented:
> Why second commit does not correspond to https://github.com/arenadata/gpdb/commit/c097366900c7699f6e10fe5f29a763792d34af41?

What do you mean? The list of changes in them is the same.

RekGRpth commented:
> > Why second commit does not correspond to [c097366](https://github.com/arenadata/gpdb/commit/c097366900c7699f6e10fe5f29a763792d34af41)?
> 
> What do you mean? The list of changes in them is the same.

```diff
2c2
< index bb6fa5335f65..5c7339d9d150 100644
---
> index eab478296971..666ed5ac8b74 100644
5c5
< @@ -3,9 +3,9 @@ FROM centos:centos7 as base
---
> @@ -12,9 +12,9 @@ RUN sed -i 's/\(override_install_langs*\)/# \1/' /etc/yum.conf && \
14c14
<                     krb5-server krb5-workstation xerces-c-devel file && \
---
>                     krb5-server krb5-workstation xerces-c-devel && \
17,18c17,18
< @@ -15,7 +15,7 @@ RUN yum makecache && \
<      yum -y install apr-devel bzip2-devel expat-devel libcurl-devel && \
---
> @@ -25,7 +25,7 @@ RUN yum makecache && \
>      yum -y install apr-devel apr-util-devel bzip2-devel expat-devel libcurl-devel && \
23,25c23,25
<      yum -y install perl-IPC-Run perl-Test-Base libxslt-devel openldap-devel && \
<      yum clean all
<  
---
>      yum -y install perl-IPC-Run perl-Test-Base libxslt-devel && \
>      yum -y install libzstd-static && \
>      # The last python 2.7 pip
```

KnightMurloc commented:
in this regard, yes, they are different because the dockerfile for 6X and 7X are different. But the list of changes (removed packages) is the same.

Stolb27 commented on arenadata/Dockerfile:
```diff
@@ -1,45 +1,49 @@
-                   ant-junit autoconf bison cmake3 flex gperf indent jq libtool gcc-c++ \
-                   krb5-server krb5-workstation xerces-c-devel file && \
-    yum clean all
+RUN dnf -y install 'dnf-command(config-manager)' && dnf config-manager --set-enabled devel && dnf makecache && dnf update -y ca-certificates && \
```
Is this changes required for vault repo? Maybe combine it together to simplify revert at further?

KnightMurloc replied:
this is necessary to add a devel repository from which some packages are taken. this is not related to llvm.

Stolb27 commented on arenadata/readme.md:
````diff
@@ -22,16 +22,13 @@ We need to execute [../concourse/scripts/ic_gpdb.bash](../concourse/scripts/ic_g
 ```
 
 * we need to modify `MAKE_TEST_COMMAND` environment variable to run different suite. e.g. we may run test againt Postgres optimizer or ORCA with altering `PGOPTIONS` environment variable;
-* we need to run container as `--privileged` to run debugger inside it
````
Why do you consider this statement as redundant? Are you able to run gdb inside an unprivileged container?

KnightMurloc replied:
yes

Stolb27 replied:
Really?

```
[root@ac4d63bf750d gpadmin]# gdb -p 13079 --batch --ex 'bt 10'
ptrace: Operation not permitted.
No stack.
```

Stolb27 commented on arenadata/readme.md:
```diff
@@ -89,15 +86,10 @@ bash arenadata/scripts/run_behave_tests.bash gpstart gpstop
 
-Tests use `allure-behave` package and store allure output files in `allure-results` folder
-**NOTE** that `allure-behave` has too old a version because it is compatible with `python2`.
-Also, the allure report for each failed test has gpdb logs attached files. See `gpMgmt/test/behave_utils/arenadata/formatter.py`
```
Why did you remove attachment generation?

KnightMurloc replied:
I did not remove the generation of attachment. I just updated the description and left only the first line. it is marked as deleted since I added a dot at the end.

Stolb27 replied:
As was mentioned in the PR description:

> The attachment of gpdb logs to the allure report will be added as a separate patch.

Stolb27 reviewed (commented):
Please provide valid commit messages and squash fixup commits as far as this PR will be merged with rebase without squashing.

BenderArenadata commented:
Failed job Resource group isolation tests 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1069566

BenderArenadata commented:
Failed job Behave tests 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1069565
